ARCMWDT: compiler generates warnings about omited function declaration

**Describe the bug**
MWDT generates warnings that routine declaration of out() inside z_cbvprintf_impl() is omitted.

**note**
MWDT is Clang-based (LLVM 16.0.6) compiler

Please also mention any information which could help others to understand
the problem you're facing:
 - What target platform are you using - arc nsim platform with arcmwdt compiler

**To Reproduce**
Steps to reproduce the behavior:
1. export ZEPHYR_TOOLCHAIN_VARIANT=arcmwdt
2. west build -p -b nsim/nsim_em samples/hello_world
3. See warnings in output

**Expected behavior**
Build system generates clean build log without warnings

**Logs and console output**
``` [2/126] Generating include/generated/zephyr/version.h
-- Zephyr version: 3.7.0-rc1 (/home/user/sandbox/zephyrproject/zephyr), build: v3.7.0-rc1-7-g83958eb937d2
[16/126] Building C object zephyr/CMakeFiles/zephyr.dir/lib/os/cbprintf_complete.c.obj
/home/user/sandbox/zephyrproject/zephyr/lib/os/cbprintf_complete.c:1327:15: warning: passing arguments to a function without a prototype is deprecated in all versions of C and is not supported in C2x [-Wdeprecated-non-prototype]
 1327 |                 int rc = out((int)*sp++, ctx);
      |                             ^
/home/user/sandbox/zephyrproject/zephyr/lib/os/cbprintf_complete.c:1377:4: warning: passing arguments to a function without a prototype is deprecated in all versions of C and is not supported in C2x [-Wdeprecated-non-prototype]
 1377 |                         OUTC(*fp++);
      |                         ^
/home/user/sandbox/zephyrproject/zephyr/lib/os/cbprintf_complete.c:1354:17: note: expanded from macro 'OUTC'
 1354 |         int rc = (*out)((int)(c), ctx); \
      |                        ^
/home/user/sandbox/zephyrproject/zephyr/lib/os/cbprintf_complete.c:1586:4: warning: passing arguments to a function without a prototype is deprecated in all versions of C and is not supported in C2x [-Wdeprecated-non-prototype]
 1586 |                         OUTC('%');
      |                         ^
/home/user/sandbox/zephyrproject/zephyr/lib/os/cbprintf_complete.c:1354:17: note: expanded from macro 'OUTC'
 1354 |         int rc = (*out)((int)(c), ctx); \
      |                        ^
/home/user/sandbox/zephyrproject/zephyr/lib/os/cbprintf_complete.c:1760:7: warning: passing arguments to a function without a prototype is deprecated in all versions of C and is not supported in C2x [-Wdeprecated-non-prototype]
 1760 |                                                 OUTC(sign);
      |                                                 ^
/home/user/sandbox/zephyrproject/zephyr/lib/os/cbprintf_complete.c:1354:17: note: expanded from macro 'OUTC'
 1354 |         int rc = (*out)((int)(c), ctx); \
      |                        ^
/home/user/sandbox/zephyrproject/zephyr/lib/os/cbprintf_complete.c:1767:6: warning: passing arguments to a function without a prototype is deprecated in all versions of C and is not supported in C2x [-Wdeprecated-non-prototype]
 1767 |                                         OUTC(pad);
      |                                         ^
/home/user/sandbox/zephyrproject/zephyr/lib/os/cbprintf_complete.c:1354:17: note: expanded from macro 'OUTC'
 1354 |         int rc = (*out)((int)(c), ctx); \
      |                        ^
/home/user/sandbox/zephyrproject/zephyr/lib/os/cbprintf_complete.c:1776:4: warning: passing arguments to a function without a prototype is deprecated in all versions of C and is not supported in C2x [-Wdeprecated-non-prototype]
 1776 |                         OUTC(sign);
      |                         ^
/home/user/sandbox/zephyrproject/zephyr/lib/os/cbprintf_complete.c:1354:17: note: expanded from macro 'OUTC'
 1354 |         int rc = (*out)((int)(c), ctx); \
      |                        ^
/home/user/sandbox/zephyrproject/zephyr/lib/os/cbprintf_complete.c:1785:6: warning: passing arguments to a function without a prototype is deprecated in all versions of C and is not supported in C2x [-Wdeprecated-non-prototype]
 1785 |                                         OUTC(*cp++);
      |                                         ^
```

**Environment (please complete the following information):**
 - OS: Linux
 - Toolchain: arcmwdt
 - v3.7.0-rc1-7-g83958eb937d2

--- a/include/zephyr/sys/cbprintf.h
+++ b/include/zephyr/sys/cbprintf.h
@@ -302,6 +302,12 @@ typedef int (*cbprintf_cb)(int c, void *ctx);
 typedef int (*cbprintf_cb)(/* int c, void *ctx */);
 #endif
 
+/* Create local cbprintf_cb type to make calng-based compilers happy when handles
+ * OUTC() macro (see below). With strict rules (Wincompatible-function-pointer-types-strict)
+ * it's prohibited to pass arguments with mismatched types.
+ */
+typedef int (*cbprintf_cb_local)(int c, void *ctx);
+
 /** @brief Signature for a cbprintf multibyte callback function.
  *
  * @param buf data.

--- a/lib/os/cbprintf_complete.c
+++ b/lib/os/cbprintf_complete.c
@@ -1316,12 +1316,13 @@ static inline void store_count(const struct conversion *conv,
 }
 
 /* Outline function to emit all characters in [sp, ep). */
-static int outs(cbprintf_cb out,
+static int outs(cbprintf_cb __out,
 		void *ctx,
 		const char *sp,
 		const char *ep)
 {
 	size_t count = 0;
+	cbprintf_cb_local out = __out;
 
 	while ((sp < ep) || ((ep == NULL) && *sp)) {
 		int rc = out((int)*sp++, ctx);
@@ -1335,12 +1336,13 @@ static int outs(cbprintf_cb out,
 	return (int)count;
 }
 
-int z_cbvprintf_impl(cbprintf_cb out, void *ctx, const char *fp,
+int z_cbvprintf_impl(cbprintf_cb __out, void *ctx, const char *fp,
 		     va_list ap, uint32_t flags)
 {
 	char buf[CONVERTED_BUFLEN];
 	size_t count = 0;
 	sint_value_type sint;
+	cbprintf_cb_local out = __out;
 
 	const bool tagged_ap = (flags & Z_CBVPRINTF_PROCESS_FLAG_TAGGED_ARGS)
 			       == Z_CBVPRINTF_PROCESS_FLAG_TAGGED_ARGS;

--- a/lib/os/cbprintf_nano.c
+++ b/lib/os/cbprintf_nano.c
@@ -73,7 +73,7 @@ static inline int convert_value(uint_value_type num, unsigned int base,
  *
  * @return printed byte count if CONFIG_CBPRINTF_LIBC_SUBSTS is set
  */
-int z_cbvprintf_impl(cbprintf_cb out, void *ctx, const char *fmt,
+int z_cbvprintf_impl(cbprintf_cb __out, void *ctx, const char *fmt,
 		     va_list ap, uint32_t flags)
 {
 	size_t count = 0;
@@ -81,6 +81,7 @@ int z_cbvprintf_impl(cbprintf_cb out, void *ctx, const char *fmt,
 	char *prefix, *data;
 	int min_width, precision, data_len;
 	char padding_mode, length_mod, special;
+	cbprintf_cb_local out = __out;
 
 	const bool tagged_ap = (flags & Z_CBVPRINTF_PROCESS_FLAG_TAGGED_ARGS)
 			       == Z_CBVPRINTF_PROCESS_FLAG_TAGGED_ARGS;

